### PR TITLE
[numpy] add op random.exponential

### PR DESCRIPTION
**Pull request by Yiyan66**
## Description ##
add op random.exponential

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/random/np_exponential_op.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file np_exponential.cc
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.cc]: should be consistent with the actual file name.

File: src/operator/numpy/random/np_exponential_op.h [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+#include "../../mxnet_op.h"
+#include "../../operator_common.h"
+#include "../../tensor/elemwise_binary_broadcast_op.h"
+#include "./dist_common.h"
+
+namespace mxnet {
+namespace op {
+
+struct NumpyExponentialParam : public dmlc::Parameter<NumpyExponentialParam> {
+    dmlc::optional<float> scale;
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: 2-space indentation.

File: src/operator/numpy/random/np_exponential_op.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    auto idx = static_cast<index_t>(dot(coord, stride));
+    out[i] =  -scales[idx] * log(threshold[i]);
+  }
+};
+
+}  // namespace mxnet_op
+
+template <typename xpu>
+void NumpyExponentialForward(const nnvm::NodeAttrs &attrs,
+                         const OpContext &ctx,
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: alignment.

File: src/operator/numpy/random/np_exponential_op.h [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  Tensor<xpu, 1, float> indicator_device = workspace.Slice(output_len, output_len + 1);
+  float indicator_host = 1.0;
+  float *indicator_device_ptr = indicator_device.dptr_;
+  Kernel<set_zero, xpu>::Launch(s, 1, indicator_device_ptr);
+  prnd->SampleUniform(&workspace, 0.0, 1.0);
+  if (param.scale.has_value()) {
+    CHECK_GE(param.scale.value(), 0.0) << "ValueError: expect scale >= 0";
+    MSHADOW_REAL_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      Kernel<scalar_exponential_kernel<DType>, xpu>::Launch(
+                                          s, outputs[0].Size(), param.scale.value(),
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: 2-space indentation.

File: src/operator/numpy/random/np_exponential_op.h [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                                          uniform_tensor.dptr_, outputs[0].dptr<DType>());
+    });
+  } else {
+    MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+        Kernel<check_legal_scale_kernel<IType>, xpu>::Launch(
+            s, inputs[0].Size(), inputs[0].dptr<IType>(), indicator_device_ptr);
+      });
+      _copy<xpu>(s, &indicator_host, indicator_device_ptr);
+      CHECK_GE(indicator_host, 0.0)
+          << "ValueError: expect scale >= 0";
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: I think this check should fit in a single line.

File: src/operator/numpy/random/np_exponential_op.h [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    MSHADOW_REAL_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      Kernel<scalar_exponential_kernel<DType>, xpu>::Launch(
+                                          s, outputs[0].Size(), param.scale.value(),
+                                          uniform_tensor.dptr_, outputs[0].dptr<DType>());
+    });
+  } else {
+    MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+        Kernel<check_legal_scale_kernel<IType>, xpu>::Launch(
+            s, inputs[0].Size(), inputs[0].dptr<IType>(), indicator_device_ptr);
+      });
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: mis-aligned here.

File: src/operator/numpy/random/np_exponential_op.h [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+namespace mxnet {
+namespace op {
+
+struct NumpyExponentialParam : public dmlc::Parameter<NumpyExponentialParam> {
+    dmlc::optional<float> scale;
+    dmlc::optional<mxnet::Tuple<int>> size;
+    DMLC_DECLARE_PARAMETER(NumpyExponentialParam) {
+        DMLC_DECLARE_FIELD(scale)
+        .set_default(dmlc::optional<float> (1.0));
```
Comment (haojin2) [inline src/operator/numpy/random/np_exponential_op.h]: `dmlc::optional<float>(1.0)`

File: tests/python/unittest/test_numpy_op.py [hunk 8]
```diff
@@ -3367,6 +3367,32 @@ def hybrid_forward(self, F, x):
...
+            test_exponential = TestRandomExp(shape)
+            if hybridize:
+                test_exponential.hybridize()
+            np_out = _np.random.exponential(size = shape)
+            mx_out = test_exponential(np.array([1]))
+    
+    for shape in shapes:
+        mx_out = np.random.exponential(np.array([1]), shape)
+        np_out = _np.random.exponential(np.array([1]).asnumpy(), shape)
+        assert_almost_equal(mx_out.asnumpy().shape, np_out.shape)
```
Comment (xidulu) [inline tests/python/unittest/test_numpy_op.py]: You'd better test the raise cases like this: https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_numpy_op.py#L3131

**Pull request merged**